client/src/test/java/io/pravega/client/security/auth/JwtTokenProviderImplTest.java

```diff
@@ -0,0 +1,184 @@
+/**
+ * Copyright (c) Dell Inc., or its subsidiaries. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ */
+package io.pravega.client.security.auth;
+
+import io.pravega.client.stream.impl.Controller;
+import lombok.extern.slf4j.Slf4j;
+import org.junit.Test;
+
+import java.time.Instant;
+import java.util.concurrent.CompletableFuture;
+import java.util.function.Supplier;
+
+import static io.pravega.client.security.auth.JwtTestUtils.createJwtBody;
+import static io.pravega.client.security.auth.JwtTestUtils.dummyToken;
+import static org.junit.Assert.*;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.when;
+
+@Slf4j
+public class JwtTokenProviderImplTest {
+
+    Controller dummyController = mock(Controller.class);
+
+    @Test
+    public void testIsWithinThresholdForRefresh() {
+        JwtTokenProviderImpl objectUnderTest = new JwtTokenProviderImpl(
+                dummyToken(), dummyController, "somescope", "somestream");
+
+        assertFalse(objectUnderTest.isWithinRefreshThreshold(Instant.ofEpochSecond(10), Instant.ofEpochSecond(30)));
+        assertFalse(objectUnderTest.isWithinRefreshThreshold(Instant.now(), Instant.now().plusSeconds(11)));
+        assertTrue(objectUnderTest.isWithinRefreshThreshold(Instant.ofEpochSecond(50), Instant.ofEpochSecond(40)));
+        assertTrue(objectUnderTest.isWithinRefreshThreshold(Instant.ofEpochSecond(50), Instant.ofEpochSecond(55)));
+        assertTrue(objectUnderTest.isWithinRefreshThreshold(Instant.ofEpochSecond(50), Instant.ofEpochSecond(53)));
+    }
+
+    @Test
+    public void testRetrievesSameTokenPassedDuringConstruction() {
+        String token = String.format("header.%s.signature", createJwtBody(
+                JwtBody.builder().expirationTime(Instant.now().plusSeconds(10000).getEpochSecond()).build()));
+        JwtTokenProviderImpl objectUnderTest = new JwtTokenProviderImpl(
+                token, mock(Controller.class), "somescope", "somestream");
+        assertEquals(token, objectUnderTest.retrieveToken());
+    }
+
+    @Test
+    public void testExpirationTimeIsNullIfExpInBodyIsNotSet() {
+
+        // See decoded parts at https://jwt.io/.
+        //
+        // The body decodes to:
```

Comment:
These seem like the encoding logic should be here so that the string can actually be expressed like this in code rather than as a comment where it isn't clear how someone can add more tests like this.

Reply:
Sure, I've modified the code to address this and pushed the change.